File: JavaScript/test/src/bucket_lifecycle.js
// 设置存储桶生命周期
function putBucketLifecycle(assert) {
  //.cssg-snippet-body-start:[put-bucket-lifecycle]
  cos.putBucketLifecycle({
      Bucket: 'examplebucket-1250000000', /* 必须 */
      Region: 'COS_REGION',     /* 存储桶所在地域，必须字段 */
      Rules: [{
          "ID": "1",
          "Status": "Enabled",
          "Filter": {},
          "Transition": {
              "Days": "30",
              "StorageClass": "STANDARD_IA"
          }
      }],
  }, function(err, data) {
      console.log(err || data);
  });
  
  //.cssg-snippet-body-end
}

// 获取存储桶生命周期
function getBucketLifecycle(assert) {
  //.cssg-snippet-body-start:[get-bucket-lifecycle]
  cos.getBucketLifecycle({
      Bucket: 'examplebucket-1250000000', /* 必须 */
      Region: 'COS_REGION',     /* 存储桶所在地域，必须字段 */
  }, function(err, data) {
      console.log(err || data);
  });
  
  //.cssg-snippet-body-end
}

// 删除存储桶生命周期
function deleteBucketLifecycle(assert) {
  //.cssg-snippet-body-start:[delete-bucket-lifecycle]
  cos.deleteBucketLifecycle({
      Bucket: 'examplebucket-1250000000', /* 必须 */
      Region: 'COS_REGION',     /* 存储桶所在地域，必须字段 */
  }, function(err, data) {
      console.log(err || data);
  });
  
  //.cssg-snippet-body-end
}

// 设置存储桶生命周期
function putBucketLifecycleArchive(assert) {
  //.cssg-snippet-body-start:[put-bucket-lifecycle-archive]
  cos.putBucketLifecycle({
      Bucket: 'examplebucket-1250000000', /* 必须 */
      Region: 'COS_REGION',     /* 存储桶所在地域，必须字段 */
      Rules: [{
          "ID": "2",
          "Filter": {
              "Prefix": "dir/",
          },
          "Status": "Enabled",
          "Transition": {
              "Days": "90",
              "StorageClass": "ARCHIVE"
          }
      }],
  }, function(err, data) {
      console.log(err || data);
  });
  
  //.cssg-snippet-body-end
}

// 设置存储桶生命周期
function putBucketLifecycleExpired(assert) {
  //.cssg-snippet-body-start:[put-bucket-lifecycle-expired]
  cos.putBucketLifecycle({
      Bucket: 'examplebucket-1250000000', /* 必须 */
      Region: 'COS_REGION',     /* 存储桶所在地域，必须字段 */
      Rules: [{
          "ID": "3",
          "Status": "Enabled",
          "Filter": {},
          "Expiration": {
              "Days": "180"
          }
      }],
  }, function(err, data) {
      console.log(err || data);
  });
  
  //.cssg-snippet-body-end
}

// 设置存储桶生命周期
function putBucketLifecycleCleanAbort(assert) {
  //.cssg-snippet-body-start:[put-bucket-lifecycle-cleanAbort]
  cos.putBucketLifecycle({
      Bucket: 'examplebucket-1250000000', /* 必须 */
      Region: 'COS_REGION',     /* 存储桶所在地域，必须字段 */
      Rules: [{
          "ID": "4",
          "Status": "Enabled",
          "Filter": {},
          "AbortIncompleteMultipartUpload": {
              "DaysAfterInitiation": "30"
          }
      }],
  }, function(err, data) {
      console.log(err || data);
  });
  
  //.cssg-snippet-body-end
}

// 设置存储桶生命周期
function putBucketLifecycleHistoryArchive(assert) {
  //.cssg-snippet-body-start:[put-bucket-lifecycle-historyArchive]
  cos.putBucketLifecycle({
      Bucket: 'examplebucket-1250000000', /* 必须 */
      Region: 'COS_REGION',     /* 存储桶所在地域，必须字段 */
      Rules: [{
          "ID": "5",
          "Status": "Enabled",
          "Filter": {},
          "NoncurrentVersionTransition": {
              "NoncurrentDays": "30",
              "StorageClass": 'ARCHIVE'
          }
      }],
  }, function(err, data) {
      console.log(err || data);
  });
  
  //.cssg-snippet-body-end
}

//.cssg-methods-pragma

test("BucketLifecycle", async function(assert) {
  // 设置存储桶生命周期
  await putBucketLifecycle(assert)

  // 获取存储桶生命周期
  await getBucketLifecycle(assert)

  // 删除存储桶生命周期
  await deleteBucketLifecycle(assert)

  // 设置存储桶生命周期
  await putBucketLifecycleArchive(assert)

  // 设置存储桶生命周期
  await putBucketLifecycleExpired(assert)

  // 设置存储桶生命周期
  await putBucketLifecycleCleanAbort(assert)

  // 设置存储桶生命周期
  await putBucketLifecycleHistoryArchive(assert)

//.cssg-methods-pragma
})

File: Go/examples/bucket_lifecycle.go
package cos

import (
	"bytes"
	"context"
	"errors"
	"fmt"
	"io/ioutil"
	"net/http"
	"net/url"
	"os"
	"strings"
	"testing"
	"time"

	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/suite"
	"github.com/tencentyun/cos-go-sdk-v5"
)

type CosTestSuite struct {
	suite.Suite
	VariableThatShouldStartAtFive int

	// CI client
	Client *cos.Client

	// Copy source client
	CClient *cos.Client

	// test_object
	TestObject string

	// special_file_name
	SepFileName string
}

var (
	UploadID string
	PartETag string
)

// 设置存储桶生命周期
func (s *CosTestSuite) putBucketLifecycle() {
  client := s.Client
  //.cssg-snippet-body-start:[put-bucket-lifecycle]
  lc := &cos.BucketPutLifecycleOptions{
      Rules: []cos.BucketLifecycleRule{
          {
              ID:     "1234",
              Filter: &cos.BucketLifecycleFilter{Prefix: "test"},
              Status: "Enabled",
              Transition: &cos.BucketLifecycleTransition{
                  Days:         10,
                  StorageClass: "Standard",
              },
          },
          {
              ID:     "123422",
              Filter: &cos.BucketLifecycleFilter{Prefix: "gg"},
              Status: "Disabled",
              Expiration: &cos.BucketLifecycleExpiration{
                  Days: 10,
              },
          },
      },
  }
  _, err := client.Bucket.PutLifecycle(context.Background(), lc)
  if err != nil {
      panic(err)
  }
  
  //.cssg-snippet-body-end
}

// 获取存储桶生命周期
func (s *CosTestSuite) getBucketLifecycle() {
  client := s.Client
  //.cssg-snippet-body-start:[get-bucket-lifecycle]
  _, _, err := client.Bucket.GetLifecycle(context.Background())
  if err != nil {
      panic(err)
  }
  
  //.cssg-snippet-body-end
}

// 删除存储桶生命周期
func (s *CosTestSuite) deleteBucketLifecycle() {
  client := s.Client
  //.cssg-snippet-body-start:[delete-bucket-lifecycle]
  _, err := client.Bucket.DeleteLifecycle(context.Background())
  if err != nil {
      panic(err)
  }
  
  //.cssg-snippet-body-end
}

//.cssg-methods-pragma


func TestCOSTestSuite(t *testing.T) {
	suite.Run(t, new(CosTestSuite))
}

func (s *CosTestSuite) TestBucketLifecycle() {
	// 将 examplebucket-1250000000 和 ap-guangzhou 修改为真实的信息
	u, _ := url.Parse("https://examplebucket-1250000000.cos.ap-guangzhou.myqcloud.com")
	b := &cos.BaseURL{BucketURL: u}
	c := cos.NewClient(b, &http.Client{
		Transport: &cos.AuthorizationTransport{
			SecretID:  os.Getenv("COS_KEY"),
			SecretKey: os.Getenv("COS_SECRET"),
		},
	})
  s.Client = c

	// 设置存储桶生命周期
	s.putBucketLifecycle()

	// 获取存储桶生命周期
	s.getBucketLifecycle()

	// 删除存储桶生命周期
	s.deleteBucketLifecycle()

	//.cssg-methods-pragma
}
